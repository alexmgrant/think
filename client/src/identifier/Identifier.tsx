import React, { FormEvent, useEffect, useState } from 'react';

import Card from '../components/card/Card';
import Button from '../components/button/Button';
import {
  getIdentifier,
  getNextIdentifier,
  updateIdentifier,
} from '../common/ApiUtils';
import { getEventValue, curry } from '../common/Utils';

const updateStateFromApi = curry(
  (updateStateFn: Function, apiFn: Function, payload: any = undefined) => {
    return (async () => {
      const response = await apiFn(payload);
      updateStateFn(response.data);
    })();
  }
);

const handleInputValue = (updateStateFn: Function) => (event: any) => {
  const inputValue = getEventValue(event);
  updateStateFn(isNaN(+inputValue) ? inputValue : +inputValue);
};

const Identifier = () => {
  const [identifier, setIdentifier] = useState('');
  const setInputIdentifierValue = handleInputValue(setIdentifier);
  const handleSetIdentifier = updateStateFromApi(setIdentifier);
  const handleFormSubmit = (event: FormEvent) => {
    event.preventDefault();
    handleSetIdentifier(updateIdentifier, { integer: Number(identifier) });
  };

  useEffect(() => {
    handleSetIdentifier(getIdentifier);
    // eslint-disable-next-line
  }, [setIdentifier]);

  return (
    <div className="App">
      <Card>
        <h4>ðŸ‘‹ Hey!</h4>
        <form onSubmit={(event) => handleFormSubmit(event)}>
          <label>
            <p>Current identifier</p>
            <input
              type="text"
              value={identifier.toString()}
              onInput={(event) => setInputIdentifierValue(event)}
            />
          </label>
          <br />
          <Button type="submit">Update Identifier</Button>
          <br />
        </form>
        <Button onClick={() => handleSetIdentifier(getNextIdentifier)}>
          Next Identifier
        </Button>
      </Card>
    </div>
  );
};

export default Identifier;
