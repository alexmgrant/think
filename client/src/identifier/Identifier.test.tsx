import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';

import Identifier from './Identifier';

let container: Element | null = null;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container as Element);
  (container as Element).remove();
  container = null;
});

it('should render Identifier', () => {
  act(() => render(<Identifier />, container));

  expect(pretty((container as Element).innerHTML)).toMatchSnapshot();
});
