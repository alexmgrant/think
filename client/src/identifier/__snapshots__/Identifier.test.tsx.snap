// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Identifier 1`] = `
"<div class=\\"App\\">
  <section class=\\"Card\\">
    <h4>ðŸ‘‹ Hey!</h4>
    <form><label>
        <p>Current identifier</p><input type=\\"text\\" value=\\"\\">
      </label><br><button class=\\"Button\\" type=\\"submit\\">Update Identifier</button><br></form><button class=\\"Button\\" type=\\"button\\">Next Identifier</button>
  </section>
</div>"
`;
